File: backend-dialysis-record/src/main/java/com/agustin/backend_dialysis_record/dto/SessionDto.java
package com.agustin.backend_dialysis_record.dto;

import jakarta.validation.constraints.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.UUID;

@Setter @Getter
public class SessionDto {
    private UUID id;

    @NotNull(message = "date is required")
    @PastOrPresent(message = "date cannot be in the future")
    private LocalDate date;

    @NotNull(message = "hour is required")
    private LocalTime hour;

    private LocalDate clinicalDate;

    @NotNull(message = "bag is required")
    @Min(value = 0, message = "bag must be >= 0 ")
    private Integer bag;

    // libre, pero no negativa
    @NotNull(message = "concentration is required")
    @DecimalMin(value = "0.0", inclusive = true, message = "concentration must be >= 0")
    private Float concentration;

    // ml, sin máximo
    @NotNull(message = "infusion is required")
    @Min(value = 0, message = "infusion must be >= 0 (ml)")
    private Integer infusion;

    // ml, sin máximo
    @NotNull(message = "drainage is required")
    @Min(value = 0, message = "drainage must be >= 0 (ml)")
    private Integer drainage;

    private Integer partial;

    @Size(max = 500, message = "observations max length is 500")
    private String observations;
    
    private Integer severityLevel;

    // se completa solo
    @Size(max = 120, message = "patientName max length is 120")
    private String patientName;

    // se completa solo
    private UUID patientId;
}

File: frontend/lib/features/sessions/models/session_dto.dart
class SessionDto {
  final String? id;

  final String? date;
  final String? hour;
  final String? clinicalDate;

  final int? bag;

  final double? concentration;
  final int? drainage;
  final int? infusion;
  final int? partial;

  final String? observations;
  final int? severityLevel;

  final String? patientName;
  final String? patientId;

  SessionDto({
    this.id,
    this.date,
    this.hour,
    this.clinicalDate,
    this.bag,
    this.concentration,
    this.drainage,
    this.infusion,
    this.partial,
    this.observations,
    this.severityLevel,
    this.patientName,
    this.patientId,
  });

  String? get effectiveDate => clinicalDate ?? date;

  bool get isNightShift =>
      date != null && clinicalDate != null && date != clinicalDate;

  factory SessionDto.fromJson(Map<String, dynamic> json) {
    int? toInt(dynamic v) {
      if (v == null) return null;
      if (v is int) return v;
      if (v is double) return v.toInt();
      return int.tryParse(v.toString());
    }

    double? toDouble(dynamic v) {
      if (v == null) return null;
      if (v is double) return v;
      if (v is int) return v.toDouble();
      return double.tryParse(v.toString());
    }

    return SessionDto(
      id: json['id']?.toString(),
      date: json['date']?.toString(),
      hour: json['hour']?.toString(),
      clinicalDate: json['clinicalDate']?.toString(),
      bag: toInt(json['bag']),
      concentration: toDouble(json['concentration']),
      drainage: toInt(json['drainage']),
      infusion: toInt(json['infusion']),
      partial: toInt(json['partial']),
      observations: json['observations']?.toString(),
      severityLevel: toInt(json['severityLevel']),
      patientName: json['patientName']?.toString(),
      patientId: json['patientId']?.toString(),
    );
  }
}
